pkg/reconciler/v1alpha1/revision/cruds.go

```diff
@@ -135,10 +135,16 @@ func (c *Reconciler) createService(ctx context.Context, rev *v1alpha1.Revision,
 func (c *Reconciler) checkAndUpdateService(ctx context.Context, rev *v1alpha1.Revision, sf serviceFactory, service *corev1.Service) (*corev1.Service, Changed, error) {
 	logger := logging.FromContext(ctx)
 
-	desiredService := sf(rev)
-
-	// Preserve the ClusterIP field in the Service's Spec, if it has been set.
-	desiredService.Spec.ClusterIP = service.Spec.ClusterIP
+	// Note: only reconcile the spec we set.
+	rawDesiredService := sf(rev)
+	desiredService := service.DeepCopy()
+	desiredService.Spec.Type = rawDesiredService.Spec.Type
+	desiredService.Spec.Selector = rawDesiredService.Spec.Selector
+	if len(desiredService.Spec.Ports) > 0 {
+		// NodePort is randomly assigned by k8s, backfill it to the raw desired service.
+		rawDesiredService.Spec.Ports[0].NodePort = desiredService.Spec.Ports[0].NodePort
+	}
+	desiredService.Spec.Ports = rawDesiredService.Spec.Ports
```

Comment:
I'm curious if all of this is still needed with the switch to ClusterIP?

Reply:
Removed unnecessary code